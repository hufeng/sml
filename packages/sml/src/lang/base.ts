import path from 'node:path'
import { exec } from 'node:child_process'
import Builder from '../common/builder'
import {
  ActorStyleType,
  BaseAst,
  DirectionType,
  GlobalConfigType,
  PackageStyle,
} from './types'

export const noop = () => {}

// ~~~~~~~~~~ builder ~~~~~~~~~~~~~~~~~~~~~~
class ConfigBuilder {
  private config: GlobalConfigType

  constructor(config: GlobalConfigType) {
    this.config = config
  }

  /**
   * setting actor style
   * @param style
   * @returns
   */
  actorStyle(style: ActorStyleType = 'default') {
    this.config.actorStyle = style
    return this
  }

  /**
   * setting package style
   * @param style
   * @returns
   */
  packageStyle(style: PackageStyle = 'Rectangle') {
    this.config.packageStyle = style
    return this
  }

  /**
   * setting direction
   * @param direction
   */
  direction(direction: DirectionType) {
    this.config.direction = direction
  }

  theme(theme: string = 'cerulean-outline') {
    this.config.theme = theme
  }
}

// ~~~~~~~~~~ lang base ~~~~~~~~~~~~~~~~

/**
 * all lang base class
 */
export class Lang<T extends { title: string; config?: GlobalConfigType }> {
  protected meta: T

  constructor(meta: T) {
    this.meta = meta
    this.meta.config = {
      actorStyle: 'default',
      direction: 'left->right',
      packageStyle: 'Rectangle',
      theme: 'cerulean-outline',
    }
  }

  get configuration() {
    return new ConfigBuilder(this.meta.config!)
  }
}

/**
 * all Emitter base class
 */
export abstract class Emitter<T extends BaseAst> {
  protected meta: T
  protected s: Builder

  constructor(meta: T) {
    this.meta = meta
    this.s = new Builder()
  }

  abstract emitCode(): string

  protected buildConfig = (s: Builder) => {
    const { config } = this.meta
    s
      // setting theme
      .str(`!theme ${config!.theme}`)
      // setting actor style
      .ifStr(
        config!.actorStyle !== 'default',
        `skinparam actorStyle ${config!.actorStyle}`,
      )
      .str(`skinparam packageStyle ${config?.packageStyle}`)
      // setting direction
      .str(`${config!.direction.replace('->', ' to ')} direction\n`)
    return s
  }

  protected buildTheme = (s: Builder) => {
    const { config } = this.meta
    s
      // setting theme
      .str(`!theme ${config!.theme}`)
  }

  plantUML(img: string) {
    const jar = path.join(__dirname, '../../bin/plantuml-1.2022.7.jar')

    exec(
      `${this.emitCode()} | java -jar ${jar} -pipe > ${img}`,
      (err, stdout, stderr) => {
        if (err) {
          throw err
        }
        console.log(stdout)
        console.log(stderr)
      },
    )
  }
}
