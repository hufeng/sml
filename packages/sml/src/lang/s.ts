export default class S {
  private buff: Array<string | S>

  constructor() {
    this.buff = []
  }

  $s(s: string) {
    this.buff.push(s)
    return this
  }

  $fn(fn: (s: S) => void) {
    fn(this)
    return this
  }

  $if(cond: boolean, str: string) {
    if (cond) {
      this.buff.push(str)
    }
    return this
  }

  $for<T>(arr: Array<T> = [], fn: (s: S, v: T, i?: number) => void) {
    if (arr.length === 0) {
      return this
    }

    arr.forEach((v, i) => {
      fn(this, v, i)
    })
    return this
  }

  $fors<T>(arr: Array<T> = [], fn: (s: S, v: T, i?: number) => void) {
    if (arr.length === 0) {
      return this
    }

    this.buff.push('')
    // append for
    arr.forEach((v, i) => {
      fn(this, v, i)
    })
    this.buff.push('')

    return this
  }

  reset() {
    this.buff = []
    return this
  }

  toString(delimiter = '\n') {
    return this.buff.join(delimiter)
  }
}
