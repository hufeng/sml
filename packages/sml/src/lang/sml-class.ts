import { Lang } from './sml'
import { globalCollections } from './sml-global'

// ~~~~~~~~~~~~~~ basic type ~~~~~~~~~~~~~~~~~
type DataType = string
export type VisibleType = 'private' | 'protected' | 'public' | 'package_private'
type ParamType = { name: string; type: DataType }
type MethodOptional = (m: MethodType) => void
type VisibleOptional = (v: { visible: VisibleType }) => void
type abstractOptional = (a: { abstract: boolean }) => void

// ~~~~~~~~~~~~~~ container type ~~~~~~~~~~~~
type FiledType = {
  name: string
  type: DataType
  visible: VisibleType
}
type MethodType = {
  abstract: boolean
  visible: VisibleType
  name: string
  params: Array<ParamType>
  ret: DataType
}
type ClazzType = {
  name: string
  abstract: boolean
  fields: Array<FiledType>
  methods: Array<MethodType>
  extends: Array<string>
  implements: Array<string>
}
type InterfaceType = {
  name: string
  implements: Array<string>
  methods: Array<MethodType>
}
type EnumFieldTyep = {
  name: string
  value?: number | string
}
type EnumType = {
  name: string
  fields: Array<EnumFieldTyep>
}
type StructType = ClazzType
type ProtocolType = InterfaceType
export interface SmlClazzMeta {
  clazzes: Array<ClazzType>
  interfaces: Array<InterfaceType>
  enums: Array<EnumType>
  structs: Array<StructType>
  protocols: Array<ProtocolType>
}

//~~~~~~~~~~~~ builder ~~~~~~~~~~~~~~~~
class ClazzBuilder {
  constructor(private clazz: ClazzType) {}

  field(name: string, type: DataType, visible: VisibleType = 'private') {
    this.clazz.fields.push({ name, type, visible })
    return this
  }

  method(
    name: string,
    params: MethodOptional,
    ret: MethodOptional,
    ...rest: Array<VisibleOptional | abstractOptional>
  ) {
    const method = {
      abstract: false,
      visible: 'public' as VisibleType,
      name,
      params: [] as Array<ParamType>,
      ret: '' as DataType,
    }
    params(method)
    ret(method)
    rest.forEach((opt) => opt(method))
    this.clazz.methods.push(method)
    return this
  }

  extends(...names: Array<string>) {
    this.clazz.extends = [...this.clazz.extends, ...names]
    return this
  }

  implements(...names: Array<string>) {
    this.clazz.implements = [...this.clazz.implements, ...names]
    return this
  }
}

class InfBuilder {
  constructor(private interfaces: InterfaceType) {}

  method(
    name: string,
    args: MethodOptional = (m: MethodType) => (m.params = []),
    ret: MethodOptional = (m: MethodType) => (m.ret = 'void'),
  ) {
    const method = {
      name,
      visible: 'public' as VisibleType,
      abstract: false,
      params: [] as Array<ParamType>,
      ret: '' as DataType,
    }
    args(method)
    ret(method)
    this.interfaces.methods.push(method)
    return this
  }

  implements(...names: Array<string>) {
    this.interfaces.implements = [...this.interfaces.implements, ...names]
    return this
  }
}

class EnumBuilder {
  constructor(private e: EnumType) {}

  field(name: string, value?: string | number) {
    this.e.fields.push({ name, value })
    return this
  }
}

// ~~~~~~~~~~~~~ define class lang modeing ~~~~~~~~~~~~~~~~~~~
enum T {
  void = 'void',
  string = `string`,
  boolean = `boolean`,
  number = `number`,

  go_int8 = `go_int8`,
  go_int16 = `go_int16`,
  go_int32 = `go_int32`,
  go_int = `go_int`,
  go_uint = `go_uint`,
  go_byte = 'go_byte',
  go_int64 = `go_int64`,
  go_uint8 = `go_uint8`,
  go_uint16 = `go_uint16`,
  go_uint32 = `go_uint32`,
  go_uint64 = `go_uint64`,
  go_float32 = `go_float32`,
  go_float64 = `go_float64`,

  py_int = 'py_int',
  py_bool = `py_bool`,
  py_float = `py_float`,

  java_byte = `java_byte`,
  java_short = `java_short`,
  java_int = `java_int`,
  java_long = `java_long`,
  java_float = `java_float`,
  java_double = `java_double`,
  java_boolean = `java_boolean`,
  java_char = `java_char`,
}
export class SmlClazzLang extends Lang {
  private meta: SmlClazzMeta

  constructor(title: string) {
    super(title)
    this.meta = {
      clazzes: [],
      interfaces: [],
      enums: [],
      structs: [],
      protocols: [],
    }
  }

  /**
   * define abstract class
   * @param name abstract class name
   * @returns
   */
  abstractClazz(name: string) {
    const clazz = {
      name,
      abstract: true,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    }
    this.meta.clazzes.push(clazz)
    return new ClazzBuilder(clazz)
  }

  /**
   * define class
   * @param name class name
   * @returns
   */
  clazz(name: string) {
    const clazz = {
      name,
      abstract: false,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    }
    this.meta.clazzes.push(clazz)
    return new ClazzBuilder(clazz)
  }

  /**
   * define interface
   * @param name interface name
   * @returns
   */
  interface(name: string) {
    const inf = {
      name,
      implements: [],
      methods: [],
    } as InterfaceType
    this.meta.interfaces.push(inf)
    return new InfBuilder(inf)
  }

  /**
   * define enum
   * @param name enum name
   * @returns
   */
  enum(name: string) {
    const e = {
      name,
      fields: [],
    }
    this.meta.enums.push(e)
    return new EnumBuilder(e)
  }

  /**
   * define struct
   * @param name struct name
   * @returns
   */
  struct(name: string) {
    const s = {
      name,
      abstract: false,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    } as ClazzType
    this.meta.structs.push(s)
    return new ClazzBuilder(s)
  }

  /**
   * define a protocol
   * @param name protocol name
   * @returns
   */
  protocol(name: string) {
    const prot = {
      name,
      implements: [],
      methods: [],
    } as InterfaceType
    this.meta.protocols.push(prot)
    return new InfBuilder(prot)
  }

  /**
   * args optional
   * @param args
   * @returns
   */
  args(...args: Array<ParamType>) {
    return (m: MethodType) => (m.params = [...m.params, ...args])
  }

  /**
   * arg optional
   */
  arg(name: string, type: string) {
    return { name, type } as ParamType
  }

  /**
   * return value type
   * @param type
   * @returns
   */
  ret(type: string) {
    return (m: MethodType) => (m.ret = type)
  }

  /**
   * setting field or method visible
   * @param v
   * @returns
   */
  visible(v: VisibleType) {
    return (val: { visible: VisibleType }) => (val.visible = v)
  }

  /**
   * setting method whether is abstract
   * @param v
   * @returns
   */
  abstract(v: boolean) {
    return (val: { abstract: boolean }) => (val.abstract = v)
  }

  t = T
}

/**
 * ClassDiagram factory
 * @param title
 * @param fn
 * @returns
 */
export function ClassDiagram(title: string, fn: (ml: SmlClazzLang) => void) {
  const lang = new SmlClazzLang(title)
  fn(lang)
  globalCollections.add(lang)
  return lang
}
