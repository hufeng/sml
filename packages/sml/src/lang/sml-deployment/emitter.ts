import S from '../s'
import { Emitter } from '../base'
import { DeploymentLangAst, DeploymentBase, DeploymentContainer } from './types'

export class PumlDeploymentEmitter extends Emitter<DeploymentLangAst> {
  emitCode() {
    const { title, zones, components, links, vlinks, rels } = this.meta

    return this.s
      .reset()
      .$s(`@startuml ${title.replace(/ /g, '_')}`)
      .$fn(this.buildConfig)
      .$s('')
      .$fors(components, this.build)
      .$for(zones, this.buildZones)
      .$for(links, this.buildLinks)
      .$for(vlinks, this.buildVLink)
      .$for(rels, this.buildRels)
      .$s('@enduml')
      .toString()
  }

  build(s: S, v: DeploymentBase) {
    s.$s(`${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>>`)
  }

  buildZones(s: S, v: DeploymentContainer) {
    s.$s(`${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>> {`)
      .$for(v.children, (s, v) =>
        s.$s(`  ${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>>`),
      )
      .$s('}')
  }
}
