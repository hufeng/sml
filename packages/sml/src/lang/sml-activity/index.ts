import { PumlActivityEmitter } from './emitter'
import { SmlActivityLang } from './syntax'

export function SmlActivityDiagram(
  title: string,
  fn: (ml: SmlActivityLang) => void,
) {
  const ast = {
    title,
    components: [],
    links: [],
    vlinks: [],
    notes: [],
  }

  fn(new SmlActivityLang(ast))

  return { ast, emitter: new PumlActivityEmitter(ast) }
}
