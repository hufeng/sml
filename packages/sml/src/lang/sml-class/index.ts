import { PumlClassEmitter } from './emitter'
import { SmlClazzLang as SmlClassLang } from './syntax'

/**
 * ClassDiagram factory
 * @param title
 * @param fn
 * @returns
 */
export function ClassDiagram(title: string, fn: (ml: SmlClassLang) => void) {
  const ast = {
    title,
    clazzes: [],
    interfaces: [],
    enums: [],
    structs: [],
    protocols: [],
  }

  fn(new SmlClassLang(ast))

  return {
    ast,
    emitter: new PumlClassEmitter(ast),
  }
}
