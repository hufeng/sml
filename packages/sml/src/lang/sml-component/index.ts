import { PumlComponentEmitter } from './emitter'
import { SmlComponentLang } from './syntax'

export function ComponentDiagram(
  title: string,
  fn: (ml: SmlComponentLang) => void,
) {
  const ast = {
    title,

    zones: [],
    nodes: [],
    databases: [],
    clouds: [],

    components: [],
    infs: [],

    notes: [],
    vlinks: [],
    links: [],
    rels: [],
  }
  fn(new SmlComponentLang(ast))
  return { ast, emitter: new PumlComponentEmitter(ast) }
}
