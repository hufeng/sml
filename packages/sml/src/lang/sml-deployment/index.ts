import { PumlDeploymentEmitter } from './emitter'
import { SmlDeploymentLang } from './syntax'

export function DeploymentDiagram(
  title: string,
  fn: (ml: SmlDeploymentLang) => void,
) {
  const ast = {
    title,
    artifacts: [],
    clouds: [],
    components: [],
    databases: [],
    nodes: [],
    queues: [],
    stacks: [],
    actors: [],
    boundary: [],
    infs: [],
    hexagons: [],
    controls: [],
  }

  fn(new SmlDeploymentLang(ast))

  return { ast, emitter: new PumlDeploymentEmitter(ast) }
}
