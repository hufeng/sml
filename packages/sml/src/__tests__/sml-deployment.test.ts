import { DeploymentDiagram } from '../lang/sml-deployment'

describe('deployment test suite', () => {
  it('test all elements', () => {
    const { emitter } = DeploymentDiagram('hello deployment', (ml) => {
      ml.actor('Customer')
        .actor('Guest')
        .aritfact(`beehive-service`)
        .aritfact('beehive-consumer')
        .boundary(`end`)
        .cloud('Network')
        .component('beehive-registry')
        .control(`control`)
        .database('mysql')
        .interface('invoke-service')
        .node(`127.0.0.1`)
        .queue('queue')
        .stack('mem')
    })
    expect(emitter.emitCode()).toMatchSnapshot()
  })
})
