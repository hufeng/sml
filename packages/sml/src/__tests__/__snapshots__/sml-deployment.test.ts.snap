// Vitest Snapshot v1

exports[`deployment test suite > test all elements 1`] = `
"@startuml hello_deployment
!theme sketchy-outline
skinparam packageStyle Rectangle
left to right direction



actor \\"Customer\\" as c_ce26601d <<actor>>

component \\"registry\\" as z_a9205dcf <<zone>> {
  component \\"beehive-registry\\" as c_4dd942bb <<component>>
}
cloud \\"web-service\\" as z_82daadfb <<cloud>> {
  component \\"nginx\\" as c_ee434023 <<component>>
  component \\"dubbo-consumer\\" as c_fc2d8b3d <<component>>
  queue \\"request queue\\" as c_9ad4f252 <<queue>>
}
cloud \\"User-Center\\" as z_42fac19d <<cloud>> {
  collections \\"login-service\\" as c_b615bcf7 <<RpcService>>
  collections \\"registry-service\\" as c_11289165 <<RpcService>>
}
database \\"persistence\\" as z_3a6988a0 <<database>> {
  component \\"mysql\\" as c_81c3b080 <<component>>
}
c_ce26601d --> c_ee434023
c_fc2d8b3d --> c_9ad4f252
c_fc2d8b3d ..> c_4dd942bb
c_9ad4f252 ..> c_b615bcf7
c_9ad4f252 ..> c_11289165
c_b615bcf7 ..> c_81c3b080
c_11289165 ..> c_81c3b080
@enduml"
`;
